openmsupply/open-msupply
# Barcode scanning: part III

Fixes #1566, #1545

# 👩🏻‍💻 What does this PR do? 
 
If you scan a barcode, and it doesn't exist in the database, you're shown the standard add item modal. After selecting and item, and allocating packs, this change will save the barcode, along with item id and selected pack size.

It also adds QR code support - it will:
- pick out the batch from the QR code
- display the add item modal, and
- disable stock lines which do not match the batch ( but only if the batch is included in the stock lines )
- highlight the pack quantity input field for the scanned batch number

It also connects the changes to the android scanner. This doesn't allow for continuous scanning, as the camera takes over the whole screen, so you have to click scan for each item. It's also not handling the QR codes very well, so android is not as robust overall as the handheld scanner.

<img width="989" alt="Screenshot 2023-04-21 at 1 07 49 PM" src="https://user-images.githubusercontent.com/9192912/233528424-9d2647c0-bf0a-4cee-83cb-ea649540e663.png">


# 🧪 How has/should this change been tested? 

- connect a scanner
- start electron
- open an outbound shipment
- click the scan button
- scan something

You'll likely not see a match yet - so there will be a warning telling you this. Enter a pack quantity, select an item and click ok. Scan this barcode again - you'll get the add item modal with the same item selected. Try in another outbound - it should match the item.

When an item and batch are matched, the other rows are disabled and the remaining pack quantity field has focus.

Try scanning a QR code with batch info - both with and without a stockline that has this batch number ( you may want to enter directly to the database for that. 

Try with no scanner attached.

Try android.

Try the web version - should behave as normal.

## 💌 Any notes for the reviewer?


## 📃 Documentation


- [ ] The barcode scanning for outbounds
